mlir/lib/Dialect/Tensor/IR/TensorOps.cpp

```diff
@@ -3793,6 +3794,30 @@ struct FoldConsecutiveConstantPadding : public OpRewritePattern<tensor::PadOp> {
 
 } // namespace
 
+LogicalResult
+PadOp::reifyResultShapes(OpBuilder &b,
+                         ReifiedRankedShapedTypeDims &reifiedReturnShapes) {
+  reifiedReturnShapes.resize(1, SmallVector<OpFoldResult>(getType().getRank()));
+  SmallVector<OpFoldResult> lp = getMixedLowPad();
+  SmallVector<OpFoldResult> hp = getMixedHighPad();
+  for (int64_t i = 0; i < getResultType().getRank(); ++i) {
+    if (!getType().isDynamicDim(i)) {
+      reifiedReturnShapes[0][i] = b.getIndexAttr(getType().getDimSize(i));
+      continue;
+    }
+    Location loc = getLoc();
+    Value dim = b.createOrFold<tensor::DimOp>(
+        loc, getSource(), b.create<arith::ConstantIndexOp>(loc, i));
+
+    affine::AffineBuilder ab(b, loc);
+    AffineExpr d0, d1, d2;
+    bindDims(b.getContext(), d0, d1, d2);
+    reifiedReturnShapes[0][i] = affine::makeComposedFoldedAffineApply(
```

Comment:
`reifyResultShapes` must return an SSA if and only if the respective dimension is dynamic. Wrap this in `getValueOrCreateConstantIntOp`.

See documentation of `reifyResultShapes`:

```
        In case a dimension in the type is static, the
        corresponding entry is an IntegerAttr. Otherwise, it is a Value. The
        given builder may be used to insert ops that compute result shapes.
```

Reply:
Isn't this already what the makeComposedFolded is doing?
I am not sure I am seeing what to do, could you post a pseudo-example where things are not done properly?

Reply:
note I fixed a bug that if this is the intended behavior, seemed to go against:
```reifyResultShapes must return an SSA if and only if the respective dimension is dynamic. Wrap this in getValueOrCreateConstantIntOp``` 

I still think the iff is the right way and I believe this is what I am doing but I may be missing something.

Reply:
The implementation of `OpFoldResult mlir::affine::makeComposedFoldedAffineApply` tries to fold the result. I.e., this may actually return an Attribute.

Reply:
We sync'ed offline with @matthias-springer turns out the reify evolved to OpFoldResult in the meantime so this is fine. I'll update the doc

Reply:
for context: https://github.com/llvm/llvm-project/commit/092372da15e5165be14cdbb7cac3cf4976fd82d0#diff-8cc8879b3d74154e46c94a24f4f1650e9233547350fc425092f8317ba63a6b62L55